presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java

```diff
@@ -55,19 +57,32 @@ public void onAfterClass(ITestClass testClass)
     @Override
     public void beforeInvocation(IInvokedMethod method, ITestResult testResult)
     {
-        started.put(getName(method), System.nanoTime());
+        beginTest(getName(method));
     }
 
     @Override
     public void afterInvocation(IInvokedMethod method, ITestResult testResult)
     {
         String name = getName(method);
-        Duration duration = Duration.nanosSince(started.remove(name));
+        Duration duration = endTest(name);
         if (duration.compareTo(SINGLE_TEST_LOGGING_THRESHOLD) > 0) {
             LOG.info("Test %s took %s", name, duration);
         }
     }
 
+    private void beginTest(String name)
+    {
+        Long existingEntry = started.putIfAbsent(name, System.nanoTime());
+        checkState(existingEntry == null, "There already is a start record for test: %s", name);
```

Comment:
It would be good to add a comment that we are explicitly disabling testng test Factory this way. And the reason why we're doing that. Otherwise someone someday will wonder why the heck some check is preventing them from writing the tests the way they want.